### PR TITLE
track/close kubelet->API connections on heartbeat failure

### PR DESCRIPTION
**Pull request by liggitt**
xref #48638
xref https://github.com/kubernetes-incubator/kube-aws/issues/598

we're already typically tracking kubelet -> API connections and have the ability to force close them as part of client cert rotation. if we do that tracking unconditionally, we gain the ability to also force close connections on heartbeat failure as well. it's a big hammer (means reestablishing pod watches, etc), but so is having all your pods evicted because you didn't heartbeat.

this intentionally does minimal refactoring/extraction of the cert connection tracking transport in case we want to backport this

* first commit unconditionally sets up the connection-tracking dialer, and moves all the cert management logic inside an if-block that gets skipped if no certificate manager is provided (view with whitespace ignored to see what actually changed)
* second commit plumbs the connection-closing function to the heartbeat loop and calls it on repeated failures

follow-ups:
* consider backporting this to 1.10, 1.9, 1.8
* refactor the connection managing dialer to not be so tightly bound to the client certificate management

/sig node
/sig api-machinery

```release-note
kubelet: fix hangs in updating Node status after network interruptions/changes between the kubelet and API server
```

### CODE REVIEW

File: cmd/kubelet/app/server.go
Comment (awly) [inline cmd/kubelet/app/server.go]: This doesn't seem to do anything useful, remove  `= nil`

File: pkg/kubelet/certificate/transport.go
Comment (awly) [inline pkg/kubelet/certificate/transport.go]: Document the new return value in func comment

**Comment by liggitt** [comment]
cc @sjenning

**Review by awly — CHANGES_REQUESTED** [review]
Recommend adding a test to make sure `OnHeartbeatFailure` triggers.

File: pkg/kubelet/certificate/transport.go
Comment (liggitt) [reply pkg/kubelet/certificate/transport.go]: done

**Comment by liggitt** [comment]
> Recommend adding a test to make sure OnHeartbeatFailure triggers.

done

File: pkg/kubelet/kubelet_node_status.go
Comment (dims) [inline pkg/kubelet/kubelet_node_status.go]: Do we want to set onRepeatedHeartbeatFailure to nil or something? (once we invoke the method)

Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: No, that would mean the kubelet would hit the same issue if the network condition was encountered twice during a single process lifetime

Comment (dims) [reply pkg/kubelet/kubelet_node_status.go]: Ack thanks!

**Comment by liggitt** [comment]
looks like the self-hosted kubelet (at least in some tests) depends on constructing an insecure client to its local API server:

`
W0507 17:53:27.936373    2203 server.go:561] New kubeClient from clientConfig error: using a custom transport with TLS certificate options or the insecure flag is not allowed
`

will look into what we can do in that case

**Comment by dims** [comment]
/lgtm

(looks like we still need approvals)

**Comment by liggitt** [comment]
/hold 
For additional manual testing

**Comment by bjhaid** [comment]
I have tested this PR and it works as expected in our environment, to verify the test case works, I ran it against a 1.9.1 kubelet, the HA IP for the apiserver is:

192.168.2.218

which fronts 3 apiservers:

192.168.1.10
192.168.1.11
192.168.1.12

On one of our kubernetes nodes I set up a blackhole route (null route) which mimics the behavior of AWS rolling ELB nodes as below:

`ip ro add blackhole 192.168.2.218`

I added an `/etc/hosts` entry pointing the apiserver URL to `192.168.1.11` and restarted dnsmasq (we use dnsmasq as a caching resolver on all our kubernetes nodes). After doing the 1.9.1 kubelet started logging:

```
 Error updating node status, will retry: error getting node "worker3.foo.com": Get https://apisever.foo.com:6443/api/v1/nodes/worker3.foo.com?resourceVersion=0: net/http: request canceled (Client.Timeout exceeded while 
awaiting headers)
```

And it was registered as not ready in the API:

```
worker3.foo.com   NotReady   <none>    242d      v1.9.1
```

It took ~30 minutes before it got back to a Ready state.

I restored the host to it's original state (pre-experiment, removing `/etc/hosts/` entries and blackhole route and restarting dnsmasq) and ran the same experiment with the patched kubelet. The following was observed:

- the kubelet was never registered as NotReady
- 2 updating node error log entries and immediately `netstat` revealed the patched kubelet had updated to the new IP of the apiserver


### NB

Hostnames and ip addresses have been scrubbed, but it does not impact the tests

**Comment by liggitt** [comment]
/assign @dchen1107 @derekwaynecarr
for review/approval

/cc @mikedanese 
fyi for creative use of the client cert connection tracking dialer

**Comment by sjenning** [comment]
lgtm as well

File: pkg/kubelet/certificate/transport_test.go
Comment (jpbetz) [inline pkg/kubelet/certificate/transport_test.go]: defer call `closeAllConns` in the test to increase coverage of the code change?

Comment (liggitt) [reply pkg/kubelet/certificate/transport_test.go]: I don't think so... closing stop should already call that... the point of this test is internally triggered connection closes

**Comment by jpbetz** [comment]
lgtm, logic looks sound

**Comment by derekwaynecarr** [comment]
kubelet changes lgtm.

/approve

**Comment by liggitt** [comment]
I plan to open picks to 1.8, 1.9, and 1.10, but will hold them until this makes it through serial/soak/scale CI tests

**Pull request merged**
